### ISSUE TITLE
SageMaker training does not converge... but estimator works locally 

### ISSUE THREAD

**Issue by superprachi** [aws/sagemaker-python-sdk#281]
Please fill out the form below.

### System Information
- **Framework (e.g. TensorFlow) / Algorithm (e.g. KMeans)**: TensorFlow
- **Framework Version**: 1.8.0
- **Python Version**: 3.6
- **CPU or GPU**: problem persists on both
- **Python SDK Version**: Most recent
- **Are you using a custom image**: No

### Describe the problem
I have uploaded custom tensorflow estimator code which trains well locally, but training gets stuck in SageMaker around a loss of 0.1? Additionally SageMaker outputs an accuracy of 0.92 (it should really be able to get to 100% accuracy as it does locally). In any case, I am able to load in the checkpoints SageMaker outputs to a local machine (or EC2 instance), but always get around 50% accuracy using the same training data SageMaker used for training (there are two classes) for any checkpoint. I receive no errors anywhere, it simply does not work! I am using a tfrecords file with 100 dog and 100 cat images total. I get the same result whether the batch size is all 200 images or 64.

### Minimal repro / logs
Here is how training proceeds locally (using TensorFlow estimators but not SageMaker)
INFO:tensorflow:loss = 0.700275, step = 1
INFO:tensorflow:global_step/sec: 1.9985
INFO:tensorflow:loss = 0.680041, step = 101 (50.041 sec)
INFO:tensorflow:global_step/sec: 2.00276
INFO:tensorflow:loss = 0.549796, step = 201 (49.932 sec)
INFO:tensorflow:global_step/sec: 2.00504
INFO:tensorflow:loss = 0.275223, step = 301 (49.873 sec)
INFO:tensorflow:global_step/sec: 2.01495
INFO:tensorflow:loss = 0.160578, step = 401 (49.630 sec)
INFO:tensorflow:global_step/sec: 1.99718
INFO:tensorflow:loss = 0.0563939, step = 501 (50.069 sec)
INFO:tensorflow:global_step/sec: 2.01035
INFO:tensorflow:loss = 0.00699503, step = 601 (49.744 sec)
INFO:tensorflow:global_step/sec: 1.9977
INFO:tensorflow:loss = 0.00196352, step = 701 (50.058 sec)
INFO:tensorflow:global_step/sec: 1.99215
INFO:tensorflow:loss = 0.000622945, step = 801 (50.196 sec)
INFO:tensorflow:global_step/sec: 1.99409
INFO:tensorflow:loss = 5.12074e-05, step = 901 (50.147 sec)
... and the loss continues to go down but the accuracy is 100% by this point.

Here is how training proceeds on SageMaker (I have tried both ml.m4.xlarge and ml.p2.xlarge instances):
INFO:sagemaker:Creating training-job with name: test-2018-07-06-23-50-30-046
2018-07-06 23:54:52,629 INFO - tensorflow - loss = 0.6918888, step = 1
2018-07-06 23:56:55,584 INFO - tensorflow - global_step/sec: 0.813308
2018-07-06 23:56:55,653 INFO - tensorflow - loss = 0.6914216, step = 101 (123.024 sec)
2018-07-06 23:58:58,121 INFO - tensorflow - global_step/sec: 0.816075
2018-07-06 23:58:58,285 INFO - tensorflow - loss = 0.3610152, step = 201 (122.632 sec)
2018-07-07 00:01:00,833 INFO - tensorflow - global_step/sec: 0.814918
2018-07-07 00:01:00,983 INFO - tensorflow - loss = 0.16919604, step = 301 (122.698 sec)
2018-07-07 00:03:04,773 INFO - tensorflow - global_step/sec: 0.806843
2018-07-07 00:03:04,874 INFO - tensorflow - loss = 0.14992207, step = 401 (123.891 sec)
2018-07-07 00:05:07,330 INFO - tensorflow - global_step/sec: 0.816815
2018-07-07 00:05:07,331 INFO - tensorflow - loss = 0.14343718, step = 501 (122.457 sec)
2018-07-07 00:07:09,781 INFO - tensorflow - global_step/sec: 0.815786
2018-07-07 00:07:09,928 INFO - tensorflow - loss = 0.118523374, step = 601 (122.597 sec)
2018-07-07 00:09:12,385 INFO - tensorflow - global_step/sec: 0.815635
2018-07-07 00:09:12,489 INFO - tensorflow - loss = 0.120991394, step = 701 (122.561 sec)
2018-07-07 00:11:15,014 INFO - tensorflow - global_step/sec: 0.815469
2018-07-07 00:11:15,372 INFO - tensorflow - loss = 0.103637785, step = 801 (122.883 sec)
2018-07-07 00:13:36,894 INFO - tensorflow - global_step/sec: 0.704819
2018-07-07 00:13:37,016 INFO - tensorflow - loss = 0.07680348, step = 901 (141.644 sec)
2018-07-07 00:15:39,399 INFO - tensorflow - global_step/sec: 0.816293
2018-07-07 00:15:39,573 INFO - tensorflow - loss = 0.09403761, step = 1001 (122.557 sec)
2018-07-07 00:17:42,509 INFO - tensorflow - global_step/sec: 0.812287
2018-07-07 00:17:42,634 INFO - tensorflow - loss = 0.12804845, step = 1101 (123.061 sec)
2018-07-07 00:19:44,727 INFO - tensorflow - global_step/sec: 0.818208
2018-07-07 00:19:44,822 INFO - tensorflow - loss = 0.10165074, step = 1201 (122.188 sec)
2018-07-07 00:21:47,709 INFO - tensorflow - global_step/sec: 0.813128
2018-07-07 00:21:47,794 INFO - tensorflow - loss = 0.1402045, step = 1301 (122.972 sec)
2018-07-07 00:23:50,566 INFO - tensorflow - global_step/sec: 0.813955
2018-07-07 00:23:50,686 INFO - tensorflow - loss = 0.118429996, step = 1401 (122.892 sec)

Here is the SageMaker notebook code: 
from sagemaker.tensorflow import TensorFlow
previous_checkpoint_path = 's3://sagemaker-tutorials1/partial/'
saved_model_path = 's3://sagemaker-tutorials1/partial/model'

dogscats_classifier = TensorFlow(entry_point='classifierpartial.py',
                               role=role,
                               checkpoint_path = previous_checkpoint_path, 
                               output_path = saved_model_path,
                               base_job_name = 'test',
                               training_steps= 2000,                                  
                               evaluation_steps= 10,
                               hyperparameters={'learning_rate': 0.0001, 'keep_checkpoint_max': None, 'save_checkpoints_secs': 1000, 'throttle_secs': 1000},
                               train_instance_count=1,
                               train_instance_type='ml.m4.xlarge')

Here is the entry_point file (classifierpartial.py): 
```
import numpy as np
import os
import tensorflow as tf
from tensorflow.python.estimator.export.export import build_raw_serving_input_receiver_fn
from tensorflow.python.estimator.export.export_output import PredictOutput

INPUT_TENSOR_NAME = "inputs"
SIGNATURE_NAME = "serving_default"
params = {"learning_rate": 0.0001}

def model_fn(features, labels, mode, params):
   # Args:
    #
    # features: This is the x-arg from the input_fn.
    # labels:   This is the y-arg from the input_fn,
    #           see e.g. train_input_fn for these two.
    # mode:     Either TRAIN, EVAL, or PREDICT
    # params:   User-defined hyper-parameters, e.g. learning-rate.
    
    """This is the model definition portion"""
    inputs = tf.identity(features[INPUT_TENSOR_NAME], name = 'inputs')
    #tf.summary.image('images', inputs, max_outputs=100)
    
    # First convolutional layer.
    conv1 = tf.layers.conv2d(inputs, name='layer_conv1',
                           filters=32, kernel_size=3,
                           padding='same', activation=tf.nn.relu)
    pool1 = tf.layers.max_pooling2d(conv1, pool_size=2, strides=2)
    # Second convolutional layer.
    conv2 = tf.layers.conv2d(pool1, name='layer_conv2',
                           filters=64, kernel_size=3,
                           padding='same', activation=tf.nn.relu)
    pool2 = tf.layers.max_pooling2d(conv2, pool_size=2, strides=2)   
    # Third convolutional layer.
    conv3 = tf.layers.conv2d(pool2, name='layer_conv3',
                           filters=128, kernel_size=3,
                           padding='same', activation=tf.nn.relu)
    pool3 = tf.layers.max_pooling2d(conv3, pool_size=2, strides=2)    
    # Flatten to a 2-rank tensor.
    flat = tf.contrib.layers.flatten(pool3)
    # First fully-connected / dense layer.
    # This uses the ReLU activation function.
    dense1 = tf.layers.dense(flat, name='layer_fc1',
                          units=512, activation=tf.nn.relu)    
    # Second fully-connected / dense layer.
    # This is the last layer so it does not use an activation function.
    net = tf.layers.dense(dense1, name='layer_fc2',
                          units=1)
    # Logits output of the neural network.
    logits = net
    # Softmax output of the neural network.
    y_pred = tf.nn.sigmoid(logits, name = 'y_pred')   
    # Classification output of the neural network.
    y_pred_cls = tf.round(y_pred, name = 'y_pred_cls')
    
    
    """This defines different modes"""
    if mode == tf.estimator.ModeKeys.PREDICT:
        predictions = {'classes': y_pred_cls, 'probabilities': y_pred}
        export_outputs = {SIGNATURE_NAME: tf.estimator.export.PredictOutput(predictions)}
        return tf.estimator.EstimatorSpec(mode, predictions=predictions, export_outputs=export_outputs)
    else:
        #labels = tf.identity(labels, name = 'true_labels')
        loss = tf.losses.sigmoid_cross_entropy(labels, logits)
        optimizer = tf.train.RMSPropOptimizer(learning_rate=params["learning_rate"])


        train_op = optimizer.minimize(
            loss=loss, global_step=tf.train.get_global_step())
    
        accuracy = tf.metrics.accuracy(labels, y_pred_cls, name = "accuracy")
        metrics = {"accuracy": accuracy}
        
        """Defines summary objects for tensorboard"""
        #tf.identity(accuracy[1], name='train_accuracy')
        #tf.summary.scalar('train_accuracy', accuracy[1])
            
        spec = tf.estimator.EstimatorSpec(
            mode=mode,
            loss=loss,
            train_op=train_op,
            eval_metric_ops=metrics)
        
    return spec

def train_input_fn(training_dir, params):
    print('getting next batch')
    return tfrec_data_input_fn([training_dir+'/trainpartial.tfrecords'], 'train/image', 'train/label', batch_size = 64, buffer_size = 200, num_epochs = None)()

def eval_input_fn(training_dir, params):
    return tfrec_data_input_fn([training_dir+'/trainpartial.tfrecords'], 'train/image', 'train/label', buffer_size = 200)()

def tfrec_data_input_fn(filenames, f1, f2, batch_size=200, buffer_size = 200, num_epochs=1, shuffle=True):
    
    def _input_fn_():
        def _parse_record(tf_record):
            features = {
                f1: tf.FixedLenFeature([], dtype=tf.string),
                f2: tf.FixedLenFeature([], dtype=tf.int64)
            }
            record = tf.parse_single_example(tf_record, features)

            image_raw = tf.decode_raw(record[f1], tf.uint8)
            image_raw = tf.reshape(image_raw, shape=(100, 100, 3))
            image_raw = tf.cast(image_raw/255, tf.float32)
            #image_raw = transform_image(image_raw, 20, 10, 5, brightness=1)
            label = tf.cast(record[f2], tf.int32)

            return {'image': image_raw}, label
        
        dataset = tf.data.TFRecordDataset(filenames)
        dataset = dataset.map(_parse_record)

        if shuffle:
            dataset = dataset.shuffle(buffer_size=buffer_size)

        dataset = dataset.repeat(num_epochs)
        dataset = dataset.batch(batch_size)

        iterator = dataset.make_one_shot_iterator()
        features, labels = iterator.get_next()
        labels = tf.reshape(labels, shape = [-1,1])
        
        return {INPUT_TENSOR_NAME:features['image']}, labels
    
    return _input_fn_

def serving_input_fn(params):
    inputs = {INPUT_TENSOR_NAME: tf.placeholder(tf.float32, [None, 100, 100, 3])}
    return tf.estimator.export.ServingInputReceiver(inputs, inputs)

def _input_fn(addr):
    img = cv2.imread(addr)
    img = cv2.resize(img, (100,100), interpolation=cv2.INTER_CUBIC)
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    img = img.astype(np.float32)
    return img



```
